### PR TITLE
[DROOLS-7591] Experiment branch : migrate a new drools-lsp parser int…

### PR DESCRIPTION
**Pull request by tkobayas**
- New parser is migrated in `drools-drl/drools-drl-parser` with org.drools.drl10.parser package
- To enable the new parser, run `mvn test -PDRL10`
- At the moment, we get lots of test failures especially in `drools-model/drools-model-codegen` and `drools-test-coverage/test-compiler-integration`

https://github.com/apache/incubator-kie-drools/issues/5678

This PR is an experiment to migrate a new DRL parser in a feature branch `dev-new-parser`.

Currently, we get test failures with the new parser enabled `mvn test -PDRL10`.

Especially,

`drools-model/drools-model-codegen`
[ERROR] Tests run: 2433, Failures: 592, Errors: 19, Skipped: 9

`drools-test-coverage/test-compiler-integration`
[ERROR] Tests run: 5396, Failures: 1249, Errors: 269, Skipped: 71

We will be able to split and fix the issues.

**This PR is a replacement of https://github.com/apache/incubator-kie-drools/pull/5677 . This one preserves commit history in drools-lsp**

Following what @tarilabs did in 
https://issues.redhat.com/browse/DROOLS-7040
https://github.com/apache/incubator-kie-drools/pull/4498
Thanks!

**Please do not squash the commits when merging**

### CODE REVIEW

**Comment by tkobayas** [comment]
Please look at https://github.com/apache/incubator-kie-drools/pull/5677 for past comments/discussions.

File: drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl10/parser/Rule_with_Extends.drl [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl10/parser/Rule_with_Extends.drl]: Please eliminate these obsolete license headers. The sooner we get rid of them the less risk we forget about them.

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DRLExprParserTest.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+/**
+ * DRLExprTreeTest
+ */
+public class DRLExprParserTest {
+
+    DrlExprParser parser;
+
+    @BeforeEach
+    public void setUp() throws Exception {
+        this.parser = new Drl10ExprParser(LanguageLevelOption.DRL6);
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DRLExprParserTest.java:44]: Maybe, get rid of the language level parameter, it's a bit confusing (can be done in the next step).

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DescrDumperTest.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        ccd = new ConstraintConnectiveDescr( );
+        constraintsAndExpr = dumper.processImplicitConstraints( expr, atomicExpr, ccd, ccd.getDescrs().indexOf( atomicExpr ), null );
+        assertThat(ccd.getDescrs().get(0).toString()).isEqualTo(expectedConstraints1);
+        assertThat(ccd.getDescrs().get(1).toString()).isEqualTo(expectedConstraints2);
+        assertThat(constraintsAndExpr[1]).isEqualTo(expectedExpr);
+        assertThat(atomicExpr.getRewrittenExpression()).isEqualTo(expectedExpr);
+    }
+
+    public ConstraintConnectiveDescr parse( final String constraint ) {
+        DrlExprParser parser = new Drl10ExprParser(LanguageLevelOption.DRL6);
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DescrDumperTest.java:361]: Maybe, get rid of the language level parameter, it's a bit confusing (can be done in the next step).

File: drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java [hunk 4]
```diff
@@ -195,7 +197,7 @@ public void testBetaModifyWithAlpha() {
                 "  $p : Person(name == \"Mario\")\n" +
                 "  $s : String(this == $p.name)\n" +
                 "then\n" +
-                "  modify($p) { setName(\"Mark\") }" +
+                "  modify($p) { setName(\"Mark\") }\n" +
```
Comment (yurloc) [inline drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java]: Is this safe? If the test was passing without the newline it should keep passing with the new parser.

File: drools-test-coverage/test-compiler-integration/pom.xml [hunk 5]
```diff
@@ -178,6 +178,27 @@
...
+<!--            <artifactId>maven-surefire-plugin</artifactId>-->
+<!--            <configuration>-->
+<!--              <systemPropertyVariables>-->
+<!--                <drools.lang.level>DRL10</drools.lang.level>-->
+<!--              </systemPropertyVariables>-->
+<!--            </configuration>-->
+<!--          </plugin>-->
+<!--        </plugins>-->
+<!--      </build>-->
+<!--    </profile>-->
```
Comment (yurloc) [inline drools-test-coverage/test-compiler-integration/pom.xml]: Please no commented-out code without an explanation.

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/lang/DRL10Lexer.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Override
+    public Token nextToken() {
+        return null;
+    }
+
+    @Override
+    public List<DroolsParserException> getErrors() {
+        return null;
```
Comment (yurloc) [inline drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/lang/DRL10Lexer.java]: Can we make this more defensive by throwing an UnsupportedOperationException? Returning null is a bit ambiguous.

File: drools-drl/drools-drl-parser/pom.xml [hunk 7]
```diff
@@ -101,6 +126,26 @@
             </plugins>
         </pluginManagement>
 
+        <plugins>
+            <!-- Consider if we will generate codes with a specific profile and copy them to src/main -->
```
Comment (yurloc) [inline drools-drl/drools-drl-parser/pom.xml:130]: I don't see the benefit of committing generated files.

File: drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl10/parser/Rule_with_Extends.drl (hunk 1 above)
Comment (yurloc) [reply drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl10/parser/Rule_with_Extends.drl]: But can be deferred to the next PR.

**Review by yurloc — APPROVED** [review]
A few minor suggestions, but otherwise looks good!

**Comment by tkobayas** [comment]
@yurloc Thank you for the comments! I'll address them and update this PR soon.

File: drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl10/parser/Rule_with_Extends.drl (hunk 1 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser-tests/src/test/resources/org/drools/drl10/parser/Rule_with_Extends.drl]: Thanks, fixed.

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DRLExprParserTest.java (hunk 2 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DRLExprParserTest.java:44]: Let's discuss it later. If we will finally drop `DRL10` prefix, the `LanguageLevelOption.DRL6` sounds valid.

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DescrDumperTest.java (hunk 3 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DescrDumperTest.java:361]: Let's discuss it later. If we will finally drop `DRL10` prefix, the `LanguageLevelOption.DRL6` sounds valid.

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DRLExprParserTest.java (hunk 2 above)
Comment (yurloc) [reply drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl10/parser/DRLExprParserTest.java:44]: Good point 🙂

File: drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java (hunk 4 above)
Comment (tkobayas) [reply drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java]: Sorry, it was a leftover of my experiment. I revert it back. The issue will be handled by https://github.com/apache/incubator-kie-drools/issues/5679

File: drools-test-coverage/test-compiler-integration/pom.xml (hunk 5 above)
Comment (tkobayas) [reply drools-test-coverage/test-compiler-integration/pom.xml]: Thanks. Removed,

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/lang/DRL10Lexer.java (hunk 6 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/lang/DRL10Lexer.java]: Thanks. Yes, I changed it to UnsupportedOperationException.

File: drools-drl/drools-drl-parser/pom.xml (hunk 7 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/pom.xml:130]: Agreed. Committing generated files makes the build a little faster, but we won't do it unless there is a large benefit.

**Review by mariofusco — APPROVED** [review]

**Comment by tkobayas** [comment]
@mariofusco Could you merge this (without squash)? Thanks!

**Pull request merged**